mcache.go

```diff
@@ -27,23 +25,25 @@ func NewMessageCache(gossip, history int) *MessageCache {
 		panic(err)
 	}
 	return &MessageCache{
-		msgs:    make(map[string]*pb.Message),
+		msgs:    make(map[string]*Message),
 		peertx:  make(map[string]map[peer.ID]int),
 		history: make([][]CacheEntry, history),
 		gossip:  gossip,
-		msgID:   DefaultMsgIdFn,
+		msgID: func(msg *Message) string {
+			return DefaultMsgIdFn(msg.Message)
+		},
 	}
 }
 
 type MessageCache struct {
-	msgs    map[string]*pb.Message
+	msgs    map[string]*Message
 	peertx  map[string]map[peer.ID]int
 	history [][]CacheEntry
 	gossip  int
-	msgID   MsgIdFunction
+	msgID   func(*Message) string
```

Comment:
We can consider referencing here a `msgIdGenerator`, similarly to how it is done in tracers. Also make `SetMsgIdFn` a `SetMsgIdGen`, anyway it is getting changed.

Reply:
let's keep the func, easier testability.